proposals/pending-state.md

```diff
@@ -0,0 +1,105 @@
+# pending-state-protocol
+An open protocol for interoperable asynchronous Web Components.
+
+Document status: Draft
+
+# Background
+
+There are a number of scenarios where a web component might depend on or perform some asynchronous work. Components may lazy-load parts of their implementation or content, perform I/O in response to user input, manage long-running computation, etc.
+
+It's often desirable to communicate the state of async work up the component tree to parent components so that they can display user affordances indicating whether the UI or content is pending some async operation.
+
+Frameworks have the ability to invent custom APIs and patterns to handle this kind of cross-component communication. To enable the same with web components in a way that's interoperable between components from diffrent sources, and possibly implemented with different libraries, we need a protocol that components can implement without reliance on a common implementation.
+
+# Goals
+
+* Allow components to communicate that they have pending asynchonous work to ancestors in the DOM tree.
+* Allow components to know if there is pending asynchronous work in descendants in the DOM tree.
+* Allow components to intercept, modify, and block pending state notifications.
+* Give guidance on the type of asynchronous work that should be notified.
+* Allow for some labelling or differentiation between types of asynchronous work.
+
+# Details
+
+## Asynchronous States
+
+There are four main states that an asynchronous operation can be in:
+
+* Not-started
+* Started
+* Sucessfully completed
+* Failed
+
+## The `pending-state` event
+
+Components with pending work indicate so by firing a composed, bubbling, `pending-state` Event, with a `promise` property:
```

Comment:
Do you see `cancelable` as having any benefit to this sort of protocol? This is how I'd expect to `Allow components to intercept, modify, and block pending state notifications.` with native events. The following could be of interest here:
```
const keepDoingWork = this.dispatchEvent(new PendingStateEvent(this.doWork()));
if (!keepDoingWork) {
    this.abortPendingWork();
}
```
This could similarly be achieved (as well as non-sync work cancelling) by outlining where to hang an `AbortController` off of the Protocol event.

Reply:
That's an interesting idea. Do you know of specific use-cases for this? Do analogous framework features, like Suspense, allow this?

Reply:
Thinking about this some more, the AbortController pattern works when the initiator of the task creates and passes along the controller. Receivers of the event are not the initiators and don't necessarily have permission to abort.

Any event dispatcher can look at `event.defaultPrevented` if they want a synchronous signal to not perform work. Asynchronous cancellation would require the event to carry a cancel callback.

Reply:
Add notes about aborting and preventing defaults. I think AbortSignal from task initiators is the way to go for task cancellation, and for preventDefault() I think it should be related to the UI affordance of the task, not execution of it.

I think it's likely that a component that is itself initiating a task is doing so via some more meaningful event than a generic pending-task, and that event can have a more meaningful preventDefault().